allennlp/semparse/contexts/sql_table_context.py

```diff
@@ -137,15 +164,27 @@ def initialize_valid_actions(self) -> Dict[str, List[str]]:
     def initialize_grammar_str(self):
         grammar_str = SQL_GRAMMAR_STR
 
-        if self.tables:
+        if self.all_tables:
             column_right_hand_sides = ['"*"']
-            for table, columns in self.tables.items():
+            for table, columns in self.all_tables.items():
                 column_right_hand_sides.extend([f'("{table}" ws "." ws "{column}")' for column in columns])
             grammar_str += "\n      col_ref \t\t = " + \
                     " / ".join(sorted(column_right_hand_sides, reverse=True))
 
-            grammar_str += generate_one_of_string('table_name', sorted(list(self.tables.keys()), reverse=True))
-
+            grammar_str += generate_one_of_string('table_name', sorted(list(self.all_tables.keys()), reverse=True))
+        biexprs = []
+        if self.tables_with_strings:
+            for table, columns in self.tables_with_strings.items():
+                biexprs.extend([f'("{table}" ws "." ws "{column}" ws binaryop ws {table}_{column}_string)'
+                                for column in columns])
+                for column in columns:
+                    self.cursor.execute(f'SELECT DISTINCT {table} . {column} FROM {table}')
+                    grammar_str += generate_one_of_string(nonterminal=f'{table}_{column}_string',
+                                                          literals=[row[0] for row in self.cursor.fetchall()],
+                                                          is_string=True)
+
+        grammar_str += 'biexpr = ( col_ref ws binaryop ws value) / (value ws binaryop ws value) / ' + \
```

Comment:
How does parsimonious handle repeated definitions like this?

Reply:
And, don't you no longer want the `(col_ref ws binaryop ws value)` rule?  This is the more-general, non-constrained version, and if it's still in the grammar, we can just do an end-run around the additional constraints, right?

Reply:
parsimonious takes the last definition, so redefining it here will override the previous definition.

I kept `(col_ref ws binaryop ws value)` because we still need it for things like `flight.flight_arrival_time = ALL ( SELECT DISTINCT ...`. To keep the grammar tight, I can change this into`(col_ref ws binaryop ws string_right_hand_side)`. Then make another rule `string_right_hand_side = ("ALL" ws query ) / ...` so that it doesn't allow numbers, etc.

Reply:
Ah, I get it; you removed `string` from `value`, so you can't actually get what I was afraid of.  That works.  If you can make it even tighter by doing what you suggested, even better.